lib/cppcheck.cpp

```diff
@@ -618,26 +686,65 @@ void CppCheck::executeRules(const std::string &tokenlist, const Tokenizer &token
         int erroffset = 0;
-        pcre *re = pcre_compile(rule.pattern.c_str(),0,&error,&erroffset,nullptr);
+        pcre * const re = pcre_compile(rule.pattern.c_str(),0,&pcreCompileErrorStr,&erroffset,nullptr);
         if (!re) {
-            if (error) {
-                ErrorLogger::ErrorMessage errmsg(std::list<ErrorLogger::ErrorMessage::FileLocation>(),
-                                                 emptyString,
-                                                 Severity::error,
-                                                 error,
-                                                 "pcre_compile",
-                                                 false);
+            if (pcreCompileErrorStr) {
+                const ErrorLogger::ErrorMessage errmsg(std::list<ErrorLogger::ErrorMessage::FileLocation>(),
+                                                       emptyString,
+                                                       Severity::error,
+                                                       pcreCompileErrorStr,
+                                                       "pcre_compile",
+                                                       false);
 
                 reportErr(errmsg);
             }
             continue;
         }
 
+        // Optimize the regex, but only if PCRE_CONFIG_JIT is available
+#ifdef PCRE_CONFIG_JIT
+        const char *pcreStudyErrorStr = nullptr;
+        pcre_extra * pcreExtra = pcre_study(re, PCRE_STUDY_JIT_COMPILE, &pcreStudyErrorStr);
+        // pcre_study() returns NULL for both errors and when it can not optimize the regex.
+        // The last argument is how one checks for errors.
+        // It is NULL if everything works, and points to an error string otherwise.
+        if (pcreStudyErrorStr) {
+            const ErrorLogger::ErrorMessage errmsg(std::list<ErrorLogger::ErrorMessage::FileLocation>(),
+                                                   emptyString,
+                                                   Severity::error,
+                                                   pcreStudyErrorStr,
+                                                   "pcre_study",
+                                                   false);
+
+            reportErr(errmsg);
+            // pcre_compile() worked, but pcre_study() returned an error. Free the resources allocated by pcre_compile().
+            pcre_free(re);
+            continue;
+        }
+#else
+        pcre_extra * pcreExtra = nullptr;
+#endif
+
         int pos = 0;
         int ovector[30]= {0};
-        while (pos < (int)str.size() && 0 <= pcre_exec(re, nullptr, str.c_str(), (int)str.size(), pos, 0, ovector, 30)) {
+        while (pos < (int)str.size()) {
+            const int pcreExecRet = pcre_exec(re, pcreExtra, str.c_str(), (int)str.size(), pos, 0, ovector, 30);
+            if (pcreExecRet < 0) {
+                const std::string errorMessage = pcreErrorCodeToString(pcreExecRet);
```

Comment:
maybe the errormessage should say that `pcre_exec()` failed.

Reply:
Done 

Example output:

```bash
$ cppcheck --rule-file=rule.xml test.cpp 
Checking test.cpp ...
(error) pcre_exec failed: This error is returned when a pattern that was successfully studied using a JIT compile option is being matched, but the memory available for the just-in-time processing stack is not large enough. See the pcrejit documentation for more details (PCRE_ERROR_JIT_STACKLIMIT)
```